### DISCUSSION TITLE
Attribute for Factory methods/Constructor "extensions"

### DISCUSSION THREAD

**Discussion by dellamonica** [dotnet/csharplang#7046] [Language Ideas] [answered]
# Attribute for Factory methods/Constructor "extensions"

- [x] Proposed
- [ ] Prototype: Not Started
- [ ] Implementation: Not Started
- [ ] Specification: Not Started

There are several proposals to the C# language to allow greater flexibility in extensions to existing types ("extension everything", "roles").

In this spirit, we propose the creation of an attribute assigned to "Factory" methods. These are static methods with the return type being extended. If such a static method with the attribute is within scope, the code can use `new MyType(arg_1,... , arg_n)` and the compiler would instead invoke the static method with the matching signature.

If two factory methods have the same signature or if the parameters of the factory method match exactly those of a non-private (maybe just public | internal?) constructor of the extended type, there is a compilation error.

## Alternative

Extension methods do not use attributes, so perhaps to keep it consistent, a different approach could be taken.
The keyword 'this' is used by extension methods already, so the signature for the "constructor extension" could look like this:
```
public static this MyType FactoryIntAndStr(int arg1, string arg2)
```

## Why?

For the same reason that extension methods simplify the code at the call site and improve discoverability and it allow libraries to extend types they do not own.

**Comment by HaloFour** [comment, marked as answer]
This feels like it could be covered by extensions, although the design has not determined if/how extension constructor members would work.  But given that is the direction the team is exploring for extension members, and it will eventually supplant how extension methods are declared today, it makes sense to consider it as a part of that project vs. having a different dedicated syntax for this.

**Reply by dellamonica** [reply]
Indeed, it can and should be part of the larger effort. However, I did not find anything talking about constructors in that proposal (although there is so much stuff that I might have missed it).

**Reply by HaloFour** [reply]
It's buried in a TODO in the proposal, so there wasn't much to miss, but at least it's on the radar.

> TODO more members (static constructors, constructors)
